CSS guide update: basic shapes with shape outside (#33804)
mdn/content: files/en-us/web/css/css_shapes/basic_shapes/index.md

The screenshot below is of the Firefox Shapes Inspector showing the reference box in use when you inspect a circle created using `shape-outside: circle(50%)`. The element has 20 pixels of padding, border and margin applied. The Shapes Inspector highlights these reference boxes.

![Text wrapping around a circle floated left. The left edge of the text is circular abutting the clipped shape on the outside of the margin with the margin following the shape clipping.](shapes-reference-box.png)

When using a basic shape, the reference box used by default is the `margin-box`. You can see in the screenshot that the shape is being defined with reference to that part of the Box Model.

While the default reference box is `margin-box`, this can be controlled. To set a different box as the reference box, include the box values after your basic shape definition.

These two declarations are the same:

```css
shape-outside: circle(50%);
shape-outside: circle(50%) margin-box;
```

> For your shape to use a different reference box, include a different {{cssxref("box-edge")}} value, for example, to use the border as a reference box for our circle, set:

Author: no. the previous example is using the default. The next example changes the default.

> For your shape to use a different reference box, include a different {{cssxref("box-edge")}} value, for example, to use the border as a reference box for our circle, set:

Reviewer: Ah, right, ok. Yup, re-reading this, it makes sense.

```css
.shape {
  shape-outside: circle(50%) border-box;
}
```

> The `margin-box` will clip the shape. This means that shapes created in reference to other shapes that extend past the margin box will have the shape clipped to the margin box. This is demonstrated following basic shapes examples.

Author: ```suggestion
Shapes created that extend past the margin box will have the shape clipped to the margin box. The following basic shapes demonstrate this.
```

> The `margin-box` will clip the shape. This means that shapes created in reference to other shapes that extend past the margin box will have the shape clipped to the margin box. This is demonstrated following basic shapes examples.

Author: i didn't know what the original meant either. Will work on this too since i am not the only one perplexed.

> The `margin-box` will clip the shape. This means that shapes created in reference to other shapes that extend past the margin box will have the shape clipped to the margin box. This is demonstrated following basic shapes examples.

Reviewer: I've read your update, and it makes a lot more sense. Nice work.

For a more extensive explanation of reference boxes as they apply to CSS Shapes, see [Understanding reference boxes for CSS shapes](http://razvancaliman.com/writing/css-shapes-reference-boxes/).

## inset()

> The [`inset()`](/en-US/docs/Web/CSS/basic-shape/inset) function defines a rectangle. This may not seem very useful as floating an item, without shapes, will give you a rectangular shape around it. However, the `inset()` type enables the definition of offsets, thus pulling the content in over the shape.

Author: ```suggestion
The [`inset()`](/en-US/docs/Web/CSS/basic-shape/inset) function defines a rectangle. This may not seem very useful as floating an item, without shapes, will give you a rectangular shape around it. However, the `inset()` type enables the definition of offsets, thus pulling the wrapping text around the reduced-size rectangle, over parts of the floated element.
```

> The [`inset()`](/en-US/docs/Web/CSS/basic-shape/inset) function defines a rectangle. This may not seem very useful as floating an item, without shapes, will give you a rectangular shape around it. However, the `inset()` type enables the definition of offsets, thus pulling the content in over the shape.

Author: I didn't like the original either, but this edit isn't quite right either. will marinate.

> The [`inset()`](/en-US/docs/Web/CSS/basic-shape/inset) function defines a rectangle. This may not seem very useful as floating an item, without shapes, will give you a rectangular shape around it. However, the `inset()` type enables the definition of offsets, thus pulling the content in over the shape.

Reviewer: Yup update seems OK here too.

{{EmbedGHLiveSample("css-examples/shapes/basic-shape/inset.html", '100%', 800)}}

You can also add the box value that you wish to use as a reference box. In the example below change the reference box from `margin-box` to `border-box`, `padding-box` or `content-box` to see how the reference box is used as the starting point before offsets are calculated changes.

{{EmbedGHLiveSample("css-examples/shapes/basic-shape/inset-box.html", '100%', 800)}}

You can also create rectangles based on distances from the top and left edges of the reference box with the [`rect()`](/en-US/docs/Web/CSS/basic-shape/rect) function, or by width and height with the [`xywh()`](/en-US/docs/Web/CSS/basic-shape/xywh) function; both of these also support optional rounded corners.

## circle()

> The [`circle()`](/en-US/docs/Web/CSS/basic-shape/circle) value for `shape-outside` can accept two possible arguments. The first is the `shape-radius`.

Author: good suggestion. will do.

The second argument is a `position`, which accepts a one- or two-value [`<position>`](/en-US/docs/Web/CSS/position_value), specified the same way as {{cssxref("background-position")}}, to indicate the position of the center of the circle. If one or both values are omitted, the values default to `center`.

To create a circle, we include a single radius value, optionally followed by the keyword **at** followed by a position value. This example has a circle applied to a square item with a width and height of `210px`, plus a margin of `20px`. This gives a total width for the reference box of `250px`. The `50%` value for the `<shape-radius>` means the radius is `125px`. The position value is set to `30%`, which is `30%` from the left and at the default vertical `center`.

{{EmbedGHLiveSample("css-examples/shapes/basic-shape/circle.html", '100%', 800)}}

Play with increasing or decreasing the size of the circle by changing the size of the radius, moving the circle around with the position value, or setting a reference box as we did for `inset()`.

> As an additional example, use the keywords `top left` for position to make a quarter circle shape in the top left corner of the page. The example below uses generated content to create a quarter-circle shape for text to flow around.

Reviewer: ```suggestion
The below example combines generated content with a `circle()` function that uses the keywords `top left` for position. This creates a quarter circle shape in the top left corner of the page for text to flow around.
```

{{EmbedGHLiveSample("css-examples/shapes/basic-shape/inset.html", '100%', 800)}}

You can also add a box value  as an alternative reference box. In the example below, try changing the reference box from `margin-box` to `border-box`, `padding-box`, or `content-box` to see how the reference box used as the starting point changes before offsets are calculated.

{{EmbedGHLiveSample("css-examples/shapes/basic-shape/inset-box.html", '100%', 800)}}

You can also create rectangles based on distances from the top and left edges of the reference box with the [`rect()`](/en-US/docs/Web/CSS/basic-shape/rect) function, or by width and height with the [`xywh()`](/en-US/docs/Web/CSS/basic-shape/xywh) function; both of these also support optional rounded corners.

## circle()

> The [`circle()`](/en-US/docs/Web/CSS/basic-shape/circle) value for `shape-outside` can accept two possible arguments. The first is the `shape-radius`.

Author: ```suggestion
The [`circle()`](/en-US/docs/Web/CSS/basic-shape/circle) value for `shape-outside` can accept two possible arguments: a `<shape-radius>` defining the size and the `<position>` defining its location.
```

## ellipse()

> An ellipse is essentially a squashed circle. As such, the [`ellipse()`](/en-US/docs/Web/CSS/basic-shape/ellipse) function acts in a very similar way to `circle()`, except that we have to specify two radii, `x` and `y` in that order.

Reviewer: ```suggestion
An ellipse is a squashed circle. As such, the [`ellipse()`](/en-US/docs/Web/CSS/basic-shape/ellipse) function acts in a very similar way to `circle()` except that we have to specify two radii, `x` and `y`, in that order.
```

These may then be followed by one or two `<position>` values, as with `circle()`, to define the location of the center of the ellipse. In the example below, we have an ellipse with an `x` radius of `40%`, a `y` radius of `50%` and the `<position>` set to `left`. This means that the center of the ellipse is at the center of the left edge of the reference box. This creates a half ellipse shape around which the text will wrap. You can change these values to see how the ellipse changes.

{{EmbedGHLiveSample("css-examples/shapes/basic-shape/ellipse.html", '100%', 800)}}

The keyword values of `closest-side` and `farthest-side` are useful in creating a quick ellipse based on the size of the floated element reference box.

{{EmbedGHLiveSample("css-examples/shapes/basic-shape/ellipse-keywords.html", '100%', 800)}}

## polygon()

> Another basic shape is more complex and enables the creation of multiple-sided polygon shape with the [`polygon()`](/en-US/docs/Web/CSS/basic-shape/polygon) function. This shape accepts three or more pairs of values (a polygon must at least draw a triangle). Each space-separated pair of values, each pair separated with a comma, are coordinates drawn with reference to the reference box. Each pair of coordinates defines an edge of the polygon, with the final edge defined by the first and last set of coordinates.

Reviewer: ```suggestion
The [`polygon()`](/en-US/docs/Web/CSS/basic-shape/polygon) function is more complex, enabling the creation of multiple-sided polygon shapes. This shape accepts three or more pairs of values (a polygon must at least draw a triangle). Each space-separated pair of values is separated with a comma and represents the coordinates of a single vertex drawn relative to the reference box. Each pair of coordinates defines an edge of the polygon, with the final edge defined by the first and last set of coordinates.
```

> The example below creates a shape for text to follow using the `polygon()`. Change any of the coordinate values to see how the shape is changed.

Reviewer: ```suggestion
The example below creates a shape for text to follow using the `polygon()` function. Try changing the coordinate values to see how the shape changes.
```
